### PR TITLE
avoid one MeshRenderData upload repeat

### PR DESCRIPTION
**Pull request by zhakesi**
避免同一MeshRenderData重复上传

### CODE REVIEW

File: cocos/2d/renderer/batcher-2d.ts [hunk 1]
```diff
@@ -225,6 +225,11 @@ export class Batcher2D implements IBatcher {
     }
 
     public uploadBuffers () {
+        this._meshDataArray.forEach(rd => {
+            rd.uploadBuffers();
+        });
+        this._meshDataArray = [];
```
Comment (pandamicro) [inline cocos/2d/renderer/batcher-2d.ts]: Only reset length in reset function, don't recreate array

File: cocos/2d/renderer/batcher-2d.ts [hunk 2]
```diff
@@ -225,6 +225,11 @@ export class Batcher2D implements IBatcher {
     }
 
     public uploadBuffers () {
+        this._meshDataArray.forEach(rd => {
```
Comment (pandamicro) [inline cocos/2d/renderer/batcher-2d.ts]: This should also be wrapped into the following if check, if no batches, then no need to upload any buffer

File: cocos/2d/renderer/batcher-2d.ts [hunk 3]
```diff
@@ -471,7 +477,9 @@ export class Batcher2D implements IBatcher {
         // Previous batch using mesh buffer
         if (rd && rd.isMeshBuffer) {
             ia = rd.requestIA(this.device);
-            rd.uploadBuffers();
+            if (-1 === this._meshDataArray.indexOf(rd)) {
```
Comment (pandamicro) [inline cocos/2d/renderer/batcher-2d.ts]: Fix linter issue

**Review by pandamicro — CHANGES_REQUESTED** [review]

File: cocos/2d/renderer/batcher-2d.ts [hunk 4]
```diff
@@ -226,6 +226,11 @@ export class Batcher2D implements IBatcher {
 
     public uploadBuffers () {
         if (this._batches.length > 0) {
+            this._meshDataArray.forEach(rd => {
+                rd.uploadBuffers();
+            });
+            this._meshDataArray.length = 0;
```
Comment (pandamicro) [inline cocos/2d/renderer/batcher-2d.ts]: Remove this line, we shouldn't reset buffer array here

File: cocos/2d/renderer/batcher-2d.ts [hunk 5]
```diff
@@ -226,6 +226,11 @@ export class Batcher2D implements IBatcher {
 
     public uploadBuffers () {
         if (this._batches.length > 0) {
+            this._meshDataArray.forEach(rd => {
```
Comment (pandamicro) [inline cocos/2d/renderer/batcher-2d.ts]: Missing arrow

**Review by pandamicro — CHANGES_REQUESTED** [review]

File: cocos/2d/renderer/batcher-2d.ts [hunk 6]
```diff
@@ -453,6 +459,7 @@ export class Batcher2D implements IBatcher {
         this.finishMergeBatches();
     }
 
+    private _meshDataArray :MeshRenderData[] = [];
```
Comment (pandamicro) [inline cocos/2d/renderer/batcher-2d.ts]: Private property should be placed together with other private properites

**Comment by pandamicro** [comment]
@cocos-robot run test cases

**Comment by fish9167** [comment]
@cocos-robot run test cases

**Review by pandamicro — APPROVED** [review]

**Pull request merged**